File: naru_agent/memory/manager.py
from __future__ import annotations

import hashlib
import logging
import threading
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel

from naru_agent.llm.base import BaseLLM
from naru_agent.memory.base import MemoryItem, MemoryStore
from naru_agent.memory.prompts import (
    FACT_EXTRACTION_PROMPT,
    MEMORY_RECONCILIATION_PROMPT,
)

logger = logging.getLogger(__name__)


class ExtractedFacts(BaseModel):
    facts: list[str]


class ReconciliationAction(BaseModel):
    fact: str
    action: str  # ADD, UPDATE, DELETE, NONE
    target_id: str | None = None
    updated_text: str | None = None


class ReconciliationResult(BaseModel):
    actions: list[ReconciliationAction]


class MemoryManager:
    """mem0-inspired memory manager with LLM-driven fact extraction and reconciliation.

    Usage:
        manager = MemoryManager(llm=my_llm, store=my_store, embed_fn=my_embed)
        manager.add("user_123", messages)
        results = manager.search("user_123", "what does the user prefer?")
    """

    def __init__(
        self,
        llm: BaseLLM,
        store: MemoryStore,
        embed_fn: Any,  # Callable[[str], list[float]]
        reconciliation_top_k: int = 3,
    ):
        self.llm = llm
        self.store = store
        self.embed_fn = embed_fn
        self._top_k = reconciliation_top_k
        self._seen_hashes: set[str] = set()
        self._seen_hashes_lock = threading.Lock()

    def add(self, user_id: str, messages: list[dict[str, str]]) -> list[dict]:
        """Extract facts from messages and reconcile with existing memories."""
        conversation = self._format_messages(messages)
        content_hash = hashlib.md5(conversation.encode()).hexdigest()

        with self._seen_hashes_lock:
            if content_hash in self._seen_hashes:
                return []
            self._seen_hashes.add(content_hash)

        # Step 1: Extract facts via LLM
        facts = self._extract_facts(conversation)
        if not facts:
            return []

        # Step 2: For each fact, find similar existing memories
        existing = []
        for fact in facts:
            embedding = self.embed_fn(fact)
            similar = self.store.search(user_id, embedding, top_k=self._top_k)
            for item in similar:
                if item.id not in [e["id"] for e in existing]:
                    existing.append({"id": item.id, "content": item.content})

        # Step 3: Reconcile via LLM
        actions = self._reconcile(facts, existing)

        # Step 4: Execute actions
        results = []
        for action in actions:
            if action.action == "ADD":
                item = MemoryItem(
                    id=str(uuid.uuid4()),
                    user_id=user_id,
                    content=action.fact,
                )
                embedding = self.embed_fn(action.fact)
                self.store.add(item, embedding)
                results.append({"event": "ADD", "memory": action.fact})

            elif action.action == "UPDATE" and action.target_id is not None:
                # Map integer ID back to real UUID
                real_id = self._resolve_id(action.target_id, existing)
                if real_id:
                    text = action.updated_text or action.fact
                    embedding = self.embed_fn(text)
                    self.store.update(real_id, text, embedding)
                    results.append({"event": "UPDATE", "memory": text})

            elif action.action == "DELETE" and action.target_id is not None:
                real_id = self._resolve_id(action.target_id, existing)
                if real_id:
                    self.store.delete(real_id)
                    results.append({"event": "DELETE", "id": real_id})

        return results

    def search(self, user_id: str, query: str, top_k: int = 5) -> list[MemoryItem]:
        """Search memories by semantic similarity."""
        embedding = self.embed_fn(query)
        return self.store.search(user_id, embedding, top_k=top_k)

    def get_context_string(self, user_id: str, query: str, top_k: int = 5) -> str:
        """Return a pre-formatted context string for prompt injection."""
        memories = self.search(user_id, query, top_k)
        if not memories:
            return ""
        lines = [f"- {m.content}" for m in memories]
        return "\n".join(lines)

    def get_all(self, user_id: str) -> list[MemoryItem]:
        return self.store.get_all(user_id)

    def _extract_facts(self, conversation: str) -> list[str]:
        prompt = FACT_EXTRACTION_PROMPT.format(conversation=conversation)
        messages = [{"role": "user", "content": prompt}]
        try:
            result = self.llm.chat_structured(
                messages, response_schema=ExtractedFacts, temperature=0.0
            )
            return result.facts
        except Exception:
            logger.exception("Failed to extract facts")
            return []

    def _reconcile(
        self, new_facts: list[str], existing: list[dict]
    ) -> list[ReconciliationAction]:
        if not existing:
            return [
                ReconciliationAction(fact=f, action="ADD") for f in new_facts
            ]

        # Map UUIDs to integers to prevent LLM hallucination
        id_map = {str(i): e["id"] for i, e in enumerate(existing)}
        existing_text = "\n".join(
            f"[{i}] {e['content']}" for i, e in enumerate(existing)
        )
        new_text = "\n".join(f"- {f}" for f in new_facts)

        prompt = MEMORY_RECONCILIATION_PROMPT.format(
            existing_memories=existing_text, new_facts=new_text
        )
        messages = [{"role": "user", "content": prompt}]
        try:
            result = self.llm.chat_structured(
                messages, response_schema=ReconciliationResult, temperature=0.0
            )
            return result.actions
        except Exception:
            logger.exception("Failed to reconcile memories, adding all as new")
            return [
                ReconciliationAction(fact=f, action="ADD") for f in new_facts
            ]

    @staticmethod
    def _resolve_id(target_id: str, existing: list[dict]) -> str | None:
        try:
            idx = int(target_id)
            if 0 <= idx < len(existing):
                return existing[idx]["id"]
        except (ValueError, IndexError):
            pass
        return None

    @staticmethod
    def _format_messages(messages: list[dict[str, str]]) -> str:
        lines = []
        for m in messages:
            role = m.get("role", "user")
            content = m.get("content", "")
            lines.append(f"{role}: {content}")
        return "\n".join(lines)

File: js/src/memory/manager.ts
import { generateObject, type LanguageModel } from "ai";
import { z } from "zod";
import { v4 as uuidv4 } from "uuid";
import type { MemoryItem, MemoryStore } from "./base.js";
import type { EmbedFn } from "../types.js";

const FactExtractionSchema = z.object({
  facts: z.array(z.string()).describe("Extracted factual statements"),
});

const ReconciliationActionSchema = z.object({
  actions: z.array(
    z.object({
      fact: z.string(),
      action: z.enum(["ADD", "UPDATE", "DELETE", "NONE"]),
      targetId: z.string().nullable().optional(),
      updatedText: z.string().nullable().optional(),
    }),
  ),
});

export class MemoryManager {
  private model: LanguageModel;
  private store: MemoryStore;
  private embedFn: EmbedFn;
  private reconciliationTopK: number;

  constructor(config: {
    model: LanguageModel;
    store: MemoryStore;
    embedFn: EmbedFn;
    reconciliationTopK?: number;
  }) {
    this.model = config.model;
    this.store = config.store;
    this.embedFn = config.embedFn;
    this.reconciliationTopK = config.reconciliationTopK ?? 3;
  }

  /**
   * Extract facts from messages and reconcile with existing memories.
   */
  async add(
    userId: string,
    messages: Array<{ role: string; content: string }>,
  ): Promise<MemoryItem[]> {
    // Step 1: Extract facts
    const conversationText = messages
      .map((m) => `${m.role}: ${m.content}`)
      .join("\n");

    const { object: extracted } = await generateObject({
      model: this.model,
      schema: FactExtractionSchema,
      prompt: `Extract factual statements about the user from this conversation. Only extract concrete facts, preferences, or personal information.\n\n${conversationText}`,
    });

    if (extracted.facts.length === 0) return [];

    // Step 2: Find similar existing memories for reconciliation (parallel)
    const factEmbeddings = await this.embedFn(extracted.facts);
    const searchResults = await Promise.all(
      factEmbeddings.map((embedding) =>
        this.store.search(userId, embedding, this.reconciliationTopK),
      ),
    );

    const seenIds = new Set<string>();
    const existingMemories: MemoryItem[] = [];
    for (const results of searchResults) {
      for (const r of results) {
        if (!seenIds.has(r.id)) {
          seenIds.add(r.id);
          existingMemories.push(r);
        }
      }
    }

    // Step 3: Reconcile
    const existingStr = existingMemories
      .map((m) => `[${m.id}] ${m.content}`)
      .join("\n");

    const { object: reconciliation } = await generateObject({
      model: this.model,
      schema: ReconciliationActionSchema,
      prompt: [
        "Given new facts and existing memories, decide what to do with each fact.",
        "Actions: ADD (new fact), UPDATE (modify existing, provide targetId + updatedText), DELETE (remove outdated, provide targetId), NONE (already exists).",
        "",
        `New facts:\n${extracted.facts.map((f, i) => `${i + 1}. ${f}`).join("\n")}`,
        "",
        existingStr
          ? `Existing memories:\n${existingStr}`
          : "No existing memories.",
      ].join("\n"),
    });

    // Step 4: Execute actions (parallel)
    const added: MemoryItem[] = [];

    await Promise.all(
      reconciliation.actions.map(async (action) => {
        switch (action.action) {
          case "ADD": {
            const [embedding] = await this.embedFn([action.fact]);
            const item: MemoryItem = {
              id: uuidv4(),
              userId,
              content: action.fact,
              metadata: {},
              createdAt: new Date(),
              updatedAt: new Date(),
            };
            await this.store.add(item, embedding);
            added.push(item);
            break;
          }
          case "UPDATE": {
            if (action.targetId && action.updatedText) {
              const [embedding] = await this.embedFn([action.updatedText]);
              await this.store.update(action.targetId, action.updatedText, embedding);
            }
            break;
          }
          case "DELETE": {
            if (action.targetId) {
              await this.store.delete(action.targetId);
            }
            break;
          }
        }
      }),
    );

    return added;
  }

  async search(
    userId: string,
    query: string,
    topK = 5,
  ): Promise<MemoryItem[]> {
    const [embedding] = await this.embedFn([query]);
    return this.store.search(userId, embedding, topK);
  }

  async getContextString(
    userId: string,
    query: string,
    topK = 5,
  ): Promise<string> {
    const memories = await this.search(userId, query, topK);
    if (memories.length === 0) return "";
    return (
      "User memories:\n" +
      memories.map((m) => `- ${m.content}`).join("\n")
    );
  }
}
